opm/simulators/utils/HDF5File.cpp

```diff
@@ -76,63 +96,71 @@ HDF5File::~HDF5File()
 
 void HDF5File::write(const std::string& group,
                      const std::string& dset,
-                     const std::vector<char>& buffer)
+                     const std::vector<char>& buffer,
+                     DataSetMode mode) const
 {
-    hid_t grp;
-    if (groupExists(m_file, group)) {
-        grp = H5Gopen2(m_file, group.c_str(), H5P_DEFAULT);
+    hid_t grp = H5I_INVALID_HID;
+    std::string realGroup = group;
+    if (mode == DataSetMode::PROCESS_SPLIT) {
+        if (group != "/")
+            realGroup += '/';
+        realGroup += dset;
+    }
+
+    OPM_BEGIN_PARALLEL_TRY_CATCH();
+
+    if (groupExists(m_file, realGroup)) {
+        grp = H5Gopen2(m_file, realGroup.c_str(), H5P_DEFAULT);
     } else {
-        auto grps = split_string(group, '/');
+        auto grps = split_string(realGroup, '/');
         std::string curr;
-        for (size_t i = 0; i < grps.size()-1; ++i) {
+        for (size_t i = 0; i < grps.size(); ++i) {
+            if (grps[i].empty())
+                continue;
             curr += '/';
             curr += grps[i];
             if (!groupExists(m_file, curr)) {
                 hid_t subgrp = H5Gcreate2(m_file, curr.c_str(), 0, H5P_DEFAULT, H5P_DEFAULT);
                 if (subgrp == H5I_INVALID_HID) {
-                    throw std::runtime_error("HDF5File: Failed to create group '" + curr + "'");
+                    throw std::runtime_error("Failed to create group '" + curr + "'");
+                }
+                if (i == grps.size() - 1) {
+                    grp = subgrp;
+                } else {
+                    H5Gclose(subgrp);
                 }
-                H5Gclose(subgrp);
+            } else if (i == grps.size() - 1) {
+                grp = H5Gopen2(m_file, realGroup.c_str(), H5P_DEFAULT);
             }
         }
-        grp = H5Gcreate2(m_file, group.c_str(), 0, H5P_DEFAULT, H5P_DEFAULT);
     }
 
     if (grp == H5I_INVALID_HID) {
-        throw std::runtime_error("HDF5File: Failed to create group '" + group + "'");
-    }
-
-    hsize_t size = buffer.size();
-    hsize_t start = 0;
-
-    hid_t space = H5Screate_simple(1, &size, nullptr);
-    hid_t dataset_id = H5Dcreate2(grp, dset.c_str(), H5T_NATIVE_CHAR, space,
-                       H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
-    if (dataset_id == H5I_INVALID_HID) {
-        throw std::runtime_error("HDF5File: Trying to write already existing dataset '" + group + '/' + dset + "'");
+        throw std::runtime_error("Failed to create group '" + realGroup + "'");
```

Comment:
What happens if a reservoir engineer sees the diagnostic "Failed to create group ..."?  Are they going to associate that with HDF5 or will there be confusion?

Reply:
i was a bit torn here. the "issue" is that this now (mostly) used in outer parallel exception handling macros which adds a prefix to the message. So in practice the exception would read "Error saving serialized state: Failed to create group '..'"

Reply:
> in practice the exception would read "Error saving serialized state: Failed to create group '..'"

Ah, okay.  Then this is fine with me.  We'll merge this now and then adjust diagnostics and the overall implementation as need arises.